planner/core/plan_cache_utils.go

```diff
@@ -470,44 +470,48 @@ func (checker *limitExtractor) Enter(in ast.Node) (out ast.Node, skipChildren bo
 				}
 			}
 		}
+	case *ast.SubqueryExpr, *ast.ExistsSubqueryExpr:
+		checker.hasSubQuery = true
 	}
 	return in, false
 }
 
 // Leave implements Visitor interface.
 func (checker *limitExtractor) Leave(in ast.Node) (out ast.Node, ok bool) {
-	return in, checker.cacheable
+	return in, checker.cacheable || !checker.hasSubQuery
 }
 
 // ExtractLimitFromAst extract limit offset and count from ast for plan cache key encode
-func ExtractLimitFromAst(node ast.Node, sctx sessionctx.Context) ([]uint64, error) {
+func extractLimitAndEstimateSubQueryFromAst(node ast.Node, sctx sessionctx.Context) ([]uint64, bool, error) {
```

Comment:
How about renaming this function to `extractMatchOptsFromAST(...) (*PlanCacheMatchOpts, bool)`?

Reply:
Updated, PTAL again